src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -298,6 +371,34 @@ private static void ChangeColor(bool foreground, ConsoleColor color)
         /// <summary>Cache of the format strings for foreground/background and ConsoleColor.</summary>
         private static readonly string[,] s_fgbgAndColorStrings = new string[2, 16]; // 2 == fg vs bg, 16 == ConsoleColor values
 
+        public static bool TryGetSpecialConsoleKey(char[] givenChars, int startIndex, int endIndex, out ConsoleKey key, out int keyLength)
+        {
+            key = (ConsoleKey)0;
+            keyLength = 0;
+            int unprocessedCharCount = endIndex - startIndex + 1;
+
+            if (unprocessedCharCount < TerminalKeyInfo.Instance.MinKeyLength)
```

Comment:
We always assume that we need to first exhaust the current unprocessed buffer and guess are keys based on the available buffer, before reading anything from the stdin (even if stdin has something available) . So, if we ended up capturing a key partially, we interpret the bytes we have to whatever key we can. Does that answer your question? If not, could you share an example?